File: lib/js/examples/decode.js
'use strict';

var $$Array     = require("bs-platform/lib/js/array.js");
var Curry       = require("bs-platform/lib/js/curry.js");
var Js_dict     = require("bs-platform/lib/js/js_dict.js");
var Json_decode = require("../src/json_decode.js");

function mapJsonObjectString(f, decoder, encoder, str) {
  var json = JSON.parse(str);
  return JSON.stringify(Js_dict.map(Curry.__1(encoder), Js_dict.map(Curry.__1(f), Json_decode.dict(decoder, json))));
}

function sum(param) {
  return $$Array.fold_left((function (prim, prim$1) {
                return prim + prim$1 | 0;
              }), 0, param);
}

console.log(mapJsonObjectString(sum, (function (param) {
            return Json_decode.array(Json_decode.$$int, param);
          }), (function (prim) {
            return prim;
          }), "\n      {\n        \"foo\": [1, 2, 3],\n        \"bar\": [9, 8, 7]\n      }\n    "));

exports.mapJsonObjectString = mapJsonObjectString;
exports.sum                 = sum;
/*  Not a pure module */

File: examples/decode.ml
(* Decoding a fixed JSON data structure using Json.Decode *)
let mapJsonObjectString f decoder (encoder: int -> Js.Json.t) str =
  let json = Json.parseOrRaise str in
  Json.Decode.(dict decoder json)
    |> Js.Dict.map ((fun v -> f v) [@bs])
    |> Json.Encode.dict encoder
    |> Json.stringify

let sum =
  Array.fold_left (+) 0

(* prints `{ "foo": 6, "bar": 24 }` *)
let _ =
  Js.log @@
    (mapJsonObjectString sum Json.Decode.(array int) Json.Encode.int {|
      {
        "foo": [1, 2, 3],
        "bar": [9, 8, 7]
      }
    |})

(* Error handling *)
let _ =
  let json = {|{ "y": 42 } |} |> Json.parseOrRaise in
  match Json.Decode.(field "x" int json) with
  | x ->
    Js.log x
  | exception Json.Decode.DecodeError msg ->
    Js.log ("Error:" ^ msg)